_posts/2016-06-08-component-relative-paths-in-angular-2.md

```diff
@@ -0,0 +1,305 @@
+
+*  header.component.ts
+*  header.component.html
+*  header.component.css
+
+This practice of using external files is especially important when your HTML or CSS is non-trivial. Using external files keeps your *.ts logic files much cleaner and easier to maintain.
+
+**header.component.ts**
+{% highlight ts %}
+{% raw %}
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector   : 'contacts-header',
+  templateUrl: 'header.component.html',
+  stylesUrl  : 'header.component.css'
+})
+export class HeaderComponent implements OnInit {
+  constructor() {}
+  ngOnInit() {}
+}
+{% endraw %}
+{% endhighlight %}
+
+**header.component.css**
+{% highlight css %}
+{% raw %}
+.navbar-fixed {
+  position:fixed;
+}
+{% endraw %}
+{% endhighlight %}
+
+**header.component.html**
+{% highlight html %}
+{% raw %}
+<nav class="navbar-fixed">
+  <div class="nav-wrapper">
+    <span class="brand-logo center">Contacts</span>
+  </div>
+</nav>
+{% endraw %}
+{% endhighlight %}
+
+Now this is where it gets tricky!
+
+The URL required is relative to the application root which is usually the location of the `index.html` web page that hosts the application. So the above component must be stored in the application root. Wow, this does not scale well.
+
+*  What if we have many components all at the root level ? ( OMG! )
+*  What if my components are organized in distinct packages ?
+*  What if we want to organize our components by feature (Angular Style Guide - Best Practice)?
+
+To explore the issue, let's consider the scenario where our details component (and files) are in the `src/app/header` package. The urls used above would cause the loader to fail and the developer would see the following 404 error in the developers console:
+
+![component_url_404](https://cloud.githubusercontent.com/assets/210413/15878482/eca5cba2-2ce0-11e6-8fb4-78868bec2644.png)
+
+Are you tempted to try and debug or introspect the exception stack and determine why the file was not found: good luck with that!
+ 
+So if path to the component HTML or CSS file is not valid, the **EASY** workaround is to add absolute paths to the URLs... but don't do it:
+
+
+**header.component.ts**
+{% highlight ts %}
+{% raw %}
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector   : 'contacts-header',
+  templateUrl: 'src/app/header/header.component.html',
+  stylesUrl  : 'src/app/header/header.component.css'
+})
+export class HeaderComponent implements OnInit {
+  constructor() {}
+  ngOnInit() {}
+}
+{% endraw %}
+{% endhighlight %}
+
+I hope you agree that this is a horrible idea and band-aid solution. 
+
+* What if I move my components to other packages? I will have to update the absolute paths...
+* What if I want to reuse my components in other applications?
+* Can I not use using relative-paths in my components ?
+
+![what](https://cloud.githubusercontent.com/assets/210413/15881627/0d0bc056-2cfd-11e6-8086-a87bfbbf91e6.gif)
```

Comment:
Actually this is funny, but I don't think it adds value to the content :D 

Can we remove the gif?